062-support-tiered-storage.md

```diff
@@ -0,0 +1,92 @@
+#  Support tiered storage natively in Strimzi
+
+## Current situation
+
+Kafka introduced tiered storage feature since 3.6. The feature is in early stage and it’s nice to support tiered storage configuration natively via Strimzi.
+
+## Motivation
+
+Supporting tiered storage via Strimzi allow Strimzi users to enable tiered storage natively with Strimzi config. By allowing to config custom configuration for `RemoteStorageManager`, users are able to adopt tiered storage feature on Strimzi easily.
+
+For simplicity, the proposal doesn't add support to customize `RemoteLogMetadataManager`. If tiered storage is enabled, the default implementation class `TopicBasedRemoteLogMetadataManager` in Kafka will be used.
+
+This proposal focuses on tiered storage support with a custom plugin brought in by the user. Other proposals in the future might add some built-in support for selected storage implementations such as S3. It should also be noted that, it’s on the end-user to ensure that their tiered storage setup works with Strimzi; the operator itself is strictly responsible for pushing down the necessary config to the broker pods.
+
+## Proposal
+
+Add a `spec.kafka.tieredStorage` field in the `Kafka` CR, which would support configuration for all tiered storage related configuration.
+
+The `tieredStorage` config contains configuration for some selected configuration of tiered storage feature. Most importantly, the configuration for `RemoteStorageManager` is needed.
+
+```
+kafka:
+  tieredStorage:
+    type: custom
+    rsm.className:
+    rsm.classPath:
+    rsm.impl.prefix: rsm.config.
+```
+
+### RemoteStorageManager (RSM)
+
+The RSM config only supports custom type. The config allows users to specify the RSM class information.
+
+Because the `RemoteStorageManager` dependency is provided by Kafka plugin, the implementation could vary case by case. It’s important to allow user to pass in additional configuration values within the supplied prefix range. The custom configuration can be supplied via  `spec.kafka.config` with the predefined prefix.
+
+### RemoteLogMetadataManager (RLMM)
+
+For simplicity, the proposal doesn't add customization for RemoteLogMetadataManager. If tiered storage is enabled, the default implementation using `TopicBasedRemoteLogMetadataManager` will be used, and corresponding configuration will be appended. 
+
+Since the default RLMM will create an internal Kafka client to talk to the server, there might be some additional configuration needed, like the SSL related setting. Those config can be put in `.spec.kafka.config` directly using the `rlmm.config.` prefix. Example can be found below.
+
+## Example configuration
+
+The below config define a sample configuration for tiered storage setup, using a `custom` RSM type.
+```
+kafka:
+  tieredStorage:
+    type: custom
+    rsm.className: com.example.kafka.tiered.storage.s3.S3RemoteStorageManager
+    rsm.classPath:  /opt/kafka/plugins/tiered-storage-s3/*
```

Comment:
Isn't better to remove this configuration and simply document some standard path where the Cluster Operator will search for TS plugin artifacts? This is similar to what we have for Kafka Connect plugins. Additionally, it would be easier to add built-in support for popular TS plugins if we decide to do so.

Reply:
> Additionally, it would be easier to add built-in support for popular TS plugins if we decide to do so.

Isn't it against a hardcoded path? Having a configurable path makes it easier to:
* Have different implementations in the same image
* Having in the future a _built-in_ plugin in some path and yet allowing users to add a custom plugin to a custom path?

Reply:
> Have different implementations in the same image

Well, I think you can copy different implementations in the same custom image, like we do with Kafka Connect connectors. Am I missing something?

Something like:

```sh
COPY ./my-plugins/ /opt/kafka/plugins/tiered-storage/
```

> Having in the future a built-in plugin in some path and yet allowing users to add a custom plugin to a custom path?

They will still be able to build their custom images if needed right?

Reply:
I did not look into the implementation in Kafka to be honest. But the way this is configured suggests very different behavior. Here you specify a classpath for a single plugin. For connectors, you specify a directory that contains separate subdirectories for each plugin loaded separately. So I do not think this is the same - you do not want to mix the tiered storage plugins. And you do not mix the connectors by putting them into the directory strucuture in plugins dir.

Reply:
In theory, I think you can copy many plugins, but just specify 1 path to use at runtime. I don't see a strong need for that, because only 1 RSM class will be used at a time, and it's up to the user to decide and config all the connectivity and setup for remote tier. For now I will not include it in the proposal.

Reply:
Is it not possible that in the future Kafka will support multiple RSM's (e.g. per topic)? I haven't reviewed the latest KIPs so I'm not sure if Kafka has completely ruled out that possibility from the get go.

Reply:
This is a good idea but I didn't hear the discussion on this approach. @showuon are you aware of any such upcoming support?